engine/runtime-compiler/src/main/java/org/enso/compiler/common_logic/BuildScopeFromModuleAlgorithm.java

```diff
@@ -0,0 +1,191 @@
+package org.enso.compiler.common_logic;
+
+import org.enso.compiler.MetadataInteropHelpers;
+import org.enso.compiler.core.ir.Module;
+import org.enso.compiler.core.ir.module.scope.Definition;
+import org.enso.compiler.core.ir.module.scope.definition.Method;
+import org.enso.compiler.core.ir.module.scope.imports.Polyglot;
+import org.enso.compiler.data.BindingsMap;
+import org.enso.compiler.pass.resolve.MethodDefinitions$;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.jdk.javaapi.CollectionConverters;
+
+/**
+ * Gathers the common logic for building the ModuleScope.
+ *
+ * <p>This is done in two places:
+ *
+ * <ol>
+ *   <li>in the compiler, gathering just the types to build StaticModuleScope,
+ *   <li>in the runtime, building Truffle nodes for the interpreter.
+ * </ol>
+ *
+ * <p>The interpreter does much more than the type-checker, so currently this only gathers the
+ * general shape of the process to try to ensure that they stay in sync. In future iterations, we
+ * may try to move more of the logic to this common place.
+ */
+public abstract class BuildScopeFromModuleAlgorithm<
+    FunctionType,
+    TypeScopeReferenceType,
+    ImportExportScopeType,
+    ModuleScopeType extends
+        CommonModuleScopeShape<FunctionType, TypeScopeReferenceType, ImportExportScopeType>,
```

Comment:
Wow! One should always beware of what one asks for!

> ModuleScopeType extends
>        CommonModuleScopeShape<FunctionType, TypeScopeReferenceType, ImportExportScopeType

Showing [object algebras](http://wiki.apidesign.org/wiki/ObjectAlgebras) to a functional programmers may lead to code that I no longer understand...

Reply:
I don't think there is any need for `T1 extends CT1<....>`.

If you want to perform operation on `ModuleScopeType`, then add a `protected abstract` method that works on `ModuleScopeType`.

Avoid `extends` in the algorithm generic arguments.

Reply:
Hmmm, okay I can rewrite it to such a form. It felt to me that it's a bit easier and allows me to separate concerns - the parts providing the `ModuleScopeType` interface were separated from the main algorithm and could be shared across 2 different 'algorithm' abstract classes.

If I put these methods directly into the algorithm, I'll need to duplicate them for each one. How can I ensure I can still share them? I guess I need to create a separate class `ModuleScopeInterface<ModuleScopeType>` that will allow me to act on the `ModuleScopeType` values?

Reply:
In these three commits I have 'simplified' the algorithms by removing the `extends` and replacing the interfaces with additional abstract methods: https://github.com/enso-org/enso/pull/11399/files/6b920eb565c878c136a864604b3557c5b1fb2623..dfa200b1d13dde772c688deeb116df46d9ee2c1b

Personally, I don't like it. It adds more boilerplate (the additional abstract methods) and the removal of `CommonModuleScopeShape` severs the connection between `ModuleScope` and `StaticModuleScope` (and their builder respectively). That was added deliberately as a kind of 'scaffolding' that ensures that as much as possible of common structure is shared between these two related-but-apart classes - as a way to try to ensure that any changes to one will be reflected in the other. Now I feel like the connection is a bit more loose, less rigid, and changes to structure of `ModuleScope` may be forgotten to be reflected in `StaticModuleScope`. Of course the `CommonModuleScopeShape` was not a perfect synchronization and it was possible still to make incompatible changes, but it was a way to relate the two classes.

Anyway, if you think that the code is better that way, I'm ok to keep it.